File: ch05/ex.5.0.hs
type Queue a = ([a], [a])

empty :: Queue a
empty = ([], [])

check :: [a] -> [a] -> Queue a
check [] r = (reverse r, [])
check f  r = (f, r)

snoc :: Queue a -> a -> Queue a
snoc (f, r) x = check f (x : r)

head :: Queue a -> a
head ([], _) = error "empty queue"
head (x : _, _) = x

tail :: Queue a -> Queue a
tail ([], _) = error "empty queue"
tail (_ : f, r) = check f r

File: ch05/ex.5.0.fs
module BatchedQueue

type Queue<'a> = 'a list * 'a list

let empty<'a> : Queue<'a> = ([], [])

let check (q : Queue<'a>) : Queue<'a> =
    match q with
    | ([], r) -> (List.rev r, [])
    | _ -> q

let snoc ((f, r) : Queue<'a>, x : 'a) : Queue<'a> =
    check (f, x :: r)

let head (q : Queue<'a>) : 'a =
    match q with
    | ([], _) -> failwith "empty queue"
    | (x :: _, _) -> x

let tail (q : Queue<'a>) : Queue<'a> =
    match q with
    | ([], _) -> failwith "empty queue"
    | (_ :: f, r) -> check (f, r)
